hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCancelAirdropHandler.java

```diff
@@ -30,24 +46,86 @@
 import com.hedera.node.app.spi.workflows.TransactionHandler;
 import com.hedera.node.config.data.TokensConfig;
 import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashSet;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
 /**
  * This class contains all workflow-related functionality regarding {@link HederaFunctionality#TOKEN_CANCEL_AIRDROP}.
- * This transaction type is used to reject tokens from an account and send them back to the treasury.
+ * This transaction type is used to cancel an airdrop which is in pending state.
  */
 @Singleton
 public class TokenCancelAirdropHandler extends BaseTokenHandler implements TransactionHandler {
 
     @Inject
-    public TokenCancelAirdropHandler() {}
+    public TokenCancelAirdropHandler() {
+        // Exists for injection
+    }
 
     @Override
-    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {}
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenCancelAirdropOrThrow();
+        final var allPendingAirdrops = op.pendingAirdrops();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+
+        for (final var airdrop : allPendingAirdrops) {
+            validateTrue(airdrop.hasSenderId(), INVALID_ACCOUNT_ID);
+            verifyAndRequireKeyForSender(airdrop.senderIdOrThrow(), context, accountStore);
+        }
+    }
+
+    /**
+     * Verifies that the sender account exists and ensures the account's key is valid and required for the transaction.
+     *
+     * @param senderId The AccountID of the specified sender whose key needs to be validated.
+     * @param context The PreHandleContext providing transaction context.
+     * @param accountStore The store to access readable account information.
+     * @throws PreCheckException If the sender's account is immutable or the sender's account ID is invalid.
+     */
+    private void verifyAndRequireKeyForSender(
+            @NonNull final AccountID senderId,
+            @NonNull final PreHandleContext context,
+            @NonNull final ReadableAccountStore accountStore)
+            throws PreCheckException {
+        final var senderAccount = accountStore.getAliasedAccountById(senderId);
+        validateTruePreCheck(senderAccount != null, INVALID_ACCOUNT_ID);
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = senderAccount.key();
+        if (key == null || !isValid(key)) {
+            throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+        }
+        context.requireKey(key);
+    }
 
     @Override
-    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {}
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
```

Comment:
I don't see this check: `Verify that the pending airdrops list contains between 1 and 10 entries, inclusive`

Reply:
Added the missing check.